File: leetcode/148.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func sortList(head *ListNode) *ListNode {
    if (head == nil || head.Next == nil) {
        return head
    }   
    
    slow := head
    fast := head.Next;
    
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }
    
    half := slow.Next
    slow.Next = nil
    
    return Merge(sortList(head), sortList(half))
}

func Merge(l1 *ListNode, l2 *ListNode) *ListNode {
    var res *ListNode = nil
    var temp_head *ListNode = nil
    
    for l1 != nil && l2 != nil {
        if l1.Val < l2.Val {
            if res == nil {
                res = l1 
                temp_head = l1 
            } else {
                temp_head.Next = l1
                temp_head = temp_head.Next
            }
            l1 = l1.Next 
        } else {
            if res == nil {
                res = l2 
                temp_head = l2 
            } else {
                temp_head.Next = l2
                temp_head = temp_head.Next
            }
            l2 = l2.Next 
        }
    }
    
    for l1 != nil {
        if res == nil {
            res = l1 
            temp_head = l1 
        } else {
            temp_head.Next = l1
            temp_head = temp_head.Next
        }
        l1 = l1.Next 
    }
    
    for l2 != nil {
        if res == nil {
            res = l2 
            temp_head = l2 
        } else {
            temp_head.Next = l2
            temp_head = temp_head.Next
        }
        l2 = l2.Next 
    }
    
    return res
}

File: leetcode/148.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    // merge sort
    ListNode* sortList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head; 
        }
        ListNode* slow = head;
        ListNode* fast = head->next;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        ListNode* half = slow->next;
        slow->next = nullptr;
        
        return Merge(sortList(head), sortList(half));
    }
    
    ListNode* Merge(ListNode* l1, ListNode* l2) {
        ListNode* res = nullptr;
        ListNode* temp_head = nullptr; 
        
        while (l1 && l2) {
            if (l1->val < l2->val) {
                if (res == nullptr) {
                    res = l1;
                    temp_head = l1;
                    l1 = l1->next;
                    continue;
                }
                temp_head->next = l1; 
                temp_head = temp_head->next;
                l1 = l1->next;
            } else {
                if (res == nullptr) {
                    res = l2;
                    temp_head = l2;
                    l2 = l2->next;
                    continue;
                }
                temp_head->next = l2; 
                temp_head = temp_head->next;
                l2 = l2->next;
            }
        }
        while (l1) {
            if (res == nullptr) {
                res = l1;
                temp_head = l1;
                l1 = l1->next;
                continue;
            }
            temp_head->next = l1; 
            temp_head = temp_head->next;
            l1 = l1->next;
        }
        while (l2) {
            if (res == nullptr) {
                res = l2;
                temp_head = l2;
                l2 = l2->next;
                continue;
            }
            temp_head->next = l2; 
            temp_head = temp_head->next;
            l2 = l2->next;
        }
        return res;
    }
};
